### PR TITLE
NEW add new function "setEntity()" and better compatibility with Multicompany

### PR DESCRIPTION
**Pull request by hregis**
@eldy  important !

### CODE REVIEW

File: htdocs/comm/propal/class/propal.class.php [hunk 1]
```diff
@@ -1138,7 +1138,7 @@ public function create($user, $notrigger = 0)
 					$sql = "UPDATE ".MAIN_DB_PREFIX."propal";
 					$sql.= " SET fk_delivery_address = ".$this->fk_delivery_address;
 					$sql.= " WHERE ref = '".$this->db->escape($this->ref)."'";
-					$sql.= " AND entity = ".$conf->entity;
+					$sql.= " AND entity = ".$this->entity;
```
Comment (eldy) [inline htdocs/comm/propal/class/propal.class.php]: This means you can create a proposal on entity 2 when your are logged on entity 1 ?

Comment (hregis) [reply htdocs/comm/propal/class/propal.class.php]: @eldy 
yes that's the goal, but there are user rights in the Multicompany module. It is a request of a lot of customers

Comment (eldy) [reply htdocs/comm/propal/class/propal.class.php]: You shoud introduce a
$sql.= " AND entity IN (".getEntity($this).")";    (to define entity allowed in write, compared to getEntites to know list en entities allowed in read)
so the module will have control on what is the entity to use and will be able to check permissions are ok. The $conf->entity will be returned by default. Without a centralized part of code controlled by the module, any simple api call that set the entity will be able to bypass any permissions.

Comment (hregis) [reply htdocs/comm/propal/class/propal.class.php]: @eldy 
maybe "setEntity()" instead "getEntity" ?
(getEntity is already used)

**Comment by hregis** [comment]
@eldy 
it's ok for you (with last PR) ?

**Comment by hregis** [comment]
@eldy ping ?

File: htdocs/core/lib/functions.lib.php [hunk 2]
```diff
@@ -95,6 +95,26 @@ function getEntity($element, $shared = 1, $currentobject = null)
...
+{
+	global $conf, $mc;
+
+	if (is_object($mc))
+	{
+		return $mc->setEntity($currentobject);
+	}
+	else
+	{
+		return $conf->entity;
```
Comment (eldy) [inline htdocs/core/lib/functions.lib.php]: Here we should have
return (is_object($currentobject) && $currentobject->id > 0) ? $currentobject->entity : $conf->entity;

When currentobject is not an object or no id yet, we are in create mode, so we use $conf->entity
When currentobject is an object and there is an id already, we are in update mode, so we use $conf->entity

So when you replaced $this->entity with setEntity($this), you have same code than before.

Comment (hregis) [reply htdocs/core/lib/functions.lib.php]: @eldy yes you right, i send a fix

**Comment by hregis** [comment]
@eldy ping

**Pull request merged**